File: 0561-array-partition/0561-array-partition.py
# 2 Approaches:
# 1. Sort the array and then compare consecutive numbers, get minimum then add to maximize: O(nlog(n))
# 2. Bucket Sort: O(range of numbers)
# class Solution:
#     def arrayPairSum(self, nums: List[int]) -> int:
#         nums.sort()
#         res = 0
#         for i in range(0,len(nums),2):
#             res += nums[i]
            
#         return res

class Solution:
    def arrayPairSum(self, nums: List[int]) -> int:
        # Find the range of numbers
        min_val = min(nums)
        max_val = max(nums)
        take = True 
        
        # Create buckets for the range [min_val, max_val]
        offset = -min_val  # Offset to handle negative indices
        buckets = [0] * (max_val - min_val + 1)
        res = 0
        carry = 0
        
        # Count occurrences of each number
        for num in nums:
            buckets[num + offset] += 1
        
        # Traverse the buckets to calculate the result
        for i in range(len(buckets)):
            while buckets[i] > 0:
                if take:
                    res += i - offset  # Convert bucket index back to the original number
                take = not take  # Toggle for pairing
                buckets[i] -= 1

        
        return res

File: 0561-array-partition/0561-array-partition.java
class Solution {
    public int arrayPairSum(int[] nums) {
        int n = nums.length;
        HashMap<Integer, Integer> map = new HashMap<>();
        int min_val = nums[0];
        int max_val = nums[0];
        int res = 0;
        //Bucket Sort
        for (int i=0; i < n; i++){
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            min_val = Math.min(min_val, nums[i]);
            max_val = Math.max(max_val, nums[i]);
        }
        Boolean Flag = false;
        for (int i = min_val; i <= max_val; i++){
            if (map.containsKey(i)){
                if (Flag){
                    map.put(i, map.get(i)-1);
                }
                int freq = map.get(i);
                if (freq % 2 == 0){
                    res += (freq/2) * i;
                    Flag = false; // even
                }
                else{
                    res += (freq-1)/2 * i;
                    res += i;
                    Flag = true; //odd
                }
            }
        }
        return res;
    }
}